File: sudoq-app/sudoqapp/src/main/java/de/sudoq/controller/sudoku/SudokuController.java
/*
 * SudoQ is a Sudoku-App for Adroid Devices with Version 2.2 at least.
 * Copyright (C) 2012  Heiko Klare, Julian Geppert, Jan-Bernhard Kordaß, Jonathan Kieling, Tim Zeitz, Timo Abele
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version. 
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. 
 * You should have received a copy of the GNU General Public License along with this program; if not, see <http://www.gnu.org/licenses/>.
 */
package de.sudoq.controller.sudoku;

import android.content.Context;

import de.sudoq.R;
import de.sudoq.model.actionTree.Action;
import de.sudoq.model.actionTree.NoteActionFactory;
import de.sudoq.model.actionTree.SolveActionFactory;
import de.sudoq.model.game.Game;
import de.sudoq.model.profile.Profile;
import de.sudoq.model.profile.Statistics;
import de.sudoq.model.sudoku.Cell;

/**
 * Der SudokuController ist dafür zuständig auf Aktionen des Benutzers mit dem
 * Spielfeld zu reagieren.
 */
public class SudokuController implements AssistanceRequestListener, ActionListener {
	/** Attributes */

	/**
	 * Hält eine Referenz auf das Game, welches Daten über das aktuelle Spiel
	 * enthält
	 */
	private Game game;

	/**
	 * Die SudokuActivity.
	 */
	private SudokuActivity context;

	/** Constructors */

	/**
	 * Erstellt einen neuen SudokuController. Wirft eine
	 * IllegalArgumentException, falls null übergeben wird.
	 * 
	 * @param game
	 *            Game, auf welchem der SudokuController arbeitet
	 * @param context
	 *            der Applikationskontext
	 * @throws IllegalArgumentException
	 *             Wird geworfen, falls null übergeben wird
	 */
	public SudokuController(Game game, SudokuActivity context) {
		if (game == null || context == null) {
			throw new IllegalArgumentException("Unvalid param!");
		}
		this.game = game;
		this.context = context;
	}

	/**
	 * Debugging
	 *
	 * @throws IllegalArgumentException
	 *             Wird geworfen, falls null übergeben wird
	 */
	private void getsucc(boolean illegal){
		if (illegal)
			throw new IllegalArgumentException("tu");
	}


	/** Methods */

	/**
	 * {@inheritDoc}
	 */
	public void onRedo() {
		game.redo();
	}

	/**
	 * {@inheritDoc}
	 */
	public void onUndo() {
		game.undo();
	}

	/**
	 * {@inheritDoc}
	 */
	public void onNoteAdd(Cell cell, int value) {
		game.addAndExecute(new NoteActionFactory().createAction(value, cell));
	}

	/**
	 * {@inheritDoc}
	 */
	public void onNoteDelete(Cell cell, int value) {
		game.addAndExecute(new NoteActionFactory().createAction(value, cell)); //TODO same code as onNoteAdd why?
	}

	/**
	 * {@inheritDoc}
	 */
	public void onAddEntry(Cell cell, int value) {
		game.addAndExecute(new SolveActionFactory().createAction(value, cell));
		if (this.game.isFinished()) {
			updateStatistics();
			handleFinish(false);
		}
	}

	public void onHintAction(Action a) {
		game.addAndExecute(a);
		if (this.game.isFinished()) {
			updateStatistics();
			handleFinish(false);
		}
	}

	/**
	 * {@inheritDoc}
	 */
	public void onDeleteEntry(Cell cell) {
		game.addAndExecute(new SolveActionFactory().createAction(Cell.EMPTYVAL, cell));
	}

	/**
	 * {@inheritDoc}
	 */
	public boolean onSolveOne() {
		boolean res = this.game.solveCell();
		if (this.game.isFinished()) {
			updateStatistics();
			handleFinish(false);
		}
		return res;
	}

	/**
	 * {@inheritDoc}
	 */
	public boolean onSolveCurrent(Cell cell) {
		boolean res = this.game.solveCell(cell);
		if (this.game.isFinished()) {
			updateStatistics();
			handleFinish(false);
		}
		return res;
	}

	/**
	 * {@inheritDoc}
	 */
	public boolean onSolveAll() {

		for (Cell f : this.game.getSudoku()) {
			if (!f.isNotWrong()) {
				this.game.addAndExecute(new SolveActionFactory().createAction(Cell.EMPTYVAL, f));
			}
		}

		boolean res = game.solveAll();

		if (res)
			handleFinish(true);
		return res;
	}

	/**
	 * Zeigt einen Gewinndialog an, der fragt, ob das Spiel beendet werden soll.
	 * 
	 * @param surrendered
	 *            TODO
	 */
	private void handleFinish(boolean surrendered) {
		context.setFinished(true, surrendered);
	}

	/**
	 * Updatet die Spielerstatistik des aktuellen Profils in der App.
	 */
	private void updateStatistics() {
		switch (game.getSudoku().getComplexity()) {
			case infernal:  incrementStatistic(Statistics.playedInfernalSudokus);  break;
			case difficult: incrementStatistic(Statistics.playedDifficultSudokus); break;
			case medium:    incrementStatistic(Statistics.playedMediumSudokus);    break;
			case easy:      incrementStatistic(Statistics.playedEasySudokus);      break;
		}
		incrementStatistic(Statistics.playedSudokus);
		Profile p = Profile.Companion.getInstance(context.getDir(context.getString(R.string.path_rel_profiles), Context.MODE_PRIVATE));

		if (p.getStatistic(Statistics.fastestSolvingTime) > game.getTime()) {
			p.setStatistic(Statistics.fastestSolvingTime, game.getTime());
		}
		if (p.getStatistic(Statistics.maximumPoints) < game.getScore()) {
			p.setStatistic(Statistics.maximumPoints, game.getScore());
		}
	}

	private void incrementStatistic(Statistics s){ //TODO this should probably be in model...
		Profile p = Profile.Companion.getInstance(context.getDir(context.getString(R.string.path_rel_profiles), Context.MODE_PRIVATE));
		p.setStatistic(s,  p.getStatistic(s) + 1);
	}

}

File: sudoq-app/sudoqapp/src/main/kotlin/de/sudoq/controller/sudoku/SudokuController.kt
/*
 * SudoQ is a Sudoku-App for Android Devices with Version 2.2 at least.
 * Copyright (C) 2012  Heiko Klare, Julian Geppert, Jan-Bernhard Kordaß, Jonathan Kieling, Tim Zeitz, Timo Abele
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version. 
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. 
 * You should have received a copy of the GNU General Public License along with this program; if not, see <http://www.gnu.org/licenses/>.
 */
package de.sudoq.controller.sudoku

import android.content.Context
import de.sudoq.R
import de.sudoq.model.actionTree.Action
import de.sudoq.model.actionTree.NoteActionFactory
import de.sudoq.model.actionTree.SolveActionFactory
import de.sudoq.model.game.Game
import de.sudoq.model.profile.ProfileSingleton
import de.sudoq.model.profile.Statistics
import de.sudoq.model.sudoku.Cell
import de.sudoq.model.sudoku.complexity.Complexity
import de.sudoq.persistence.profile.ProfileRepo
import de.sudoq.persistence.profile.ProfilesListRepo

/**
 * Der SudokuController ist dafür zuständig auf Aktionen des Benutzers mit dem
 * Spielfeld zu reagieren.
 */
class SudokuController(
    /** Hält eine Referenz auf das Game, welches Daten über das aktuelle Spiel enthält */
    private val game: Game,
    /** Die SudokuActivity. */
    private val context: SudokuActivity
) : AssistanceRequestListener, ActionListener {

    /**
     * Debugging
     *
     * @throws IllegalArgumentException
     * Wird geworfen, falls null übergeben wird
     */
    private fun getsucc(illegal: Boolean) {
        require(!illegal) { "tu" }
    }
    /** Methods  */
    /**
     * {@inheritDoc}
     */
    override fun onRedo() {
        game.redo()
    }

    /**
     * {@inheritDoc}
     */
    override fun onUndo() {
        game.undo()
    }

    /**
     * {@inheritDoc}
     */
    override fun onNoteAdd(cell: Cell, value: Int) {
        game.addAndExecute(NoteActionFactory().createAction(value, cell))
    }

    /**
     * {@inheritDoc}
     */
    override fun onNoteDelete(cell: Cell, value: Int) {
        game.addAndExecute(
            NoteActionFactory().createAction(
                value,
                cell
            )
        ) //TODO same code as onNoteAdd why?
    }

    /**
     * {@inheritDoc}
     */
    override fun onAddEntry(cell: Cell, value: Int) {
        game.addAndExecute(SolveActionFactory().createAction(value, cell))
        if (game.isFinished()) {
            updateStatistics()
            handleFinish(false)
        }
    }

    fun onHintAction(a: Action) {
        game.addAndExecute(a)
        if (game.isFinished()) {
            updateStatistics()
            handleFinish(false)
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun onDeleteEntry(cell: Cell) {
        game.addAndExecute(SolveActionFactory().createAction(Cell.EMPTYVAL, cell))
    }

    /**
     * {@inheritDoc}
     */
    override fun onSolveOne(): Boolean {
        val res = game.solveCell()
        if (game.isFinished()) {
            updateStatistics()
            handleFinish(false)
        }
        return res
    }

    /**
     * {@inheritDoc}
     */
    override fun onSolveCurrent(cell: Cell): Boolean {
        val res = game.solveCell(cell)
        if (game.isFinished()) {
            updateStatistics()
            handleFinish(false)
        }
        return res
    }

    /**
     * {@inheritDoc}
     */
    override fun onSolveAll(): Boolean {
        for (f in game.sudoku!!) {
            if (!f.isNotWrong) {
                game.addAndExecute(SolveActionFactory().createAction(Cell.EMPTYVAL, f))
            }
        }
        val res = game.solveAll()
        if (res) handleFinish(true)
        return res
    }

    /**
     * Zeigt einen Gewinndialog an, der fragt, ob das Spiel beendet werden soll.
     *
     * @param surrendered
     * TODO
     */
    private fun handleFinish(surrendered: Boolean) {
        context.setFinished(true, surrendered)
    }

    /**
     * Updatet die Spielerstatistik des aktuellen Profils in der App.
     */
    private fun updateStatistics() {
        when (game.sudoku!!.complexity!!) {
            Complexity.infernal -> incrementStatistic(Statistics.playedInfernalSudokus)
            Complexity.difficult -> incrementStatistic(Statistics.playedDifficultSudokus)
            Complexity.medium -> incrementStatistic(Statistics.playedMediumSudokus)
            Complexity.easy -> incrementStatistic(Statistics.playedEasySudokus)
            Complexity.arbitrary -> throw IllegalStateException("unexpected complexity value: 'arbitrary'")
        }
        incrementStatistic(Statistics.playedSudokus)
        val profilesDir = context.getDir(
            context.getString(R.string.path_rel_profiles),
            Context.MODE_PRIVATE
        )
        val p = ProfileSingleton.getInstance(profilesDir, ProfileRepo(profilesDir),
                                             ProfilesListRepo(profilesDir))
        if (p.getStatistic(Statistics.fastestSolvingTime) > game.time) {
            p.setStatistic(Statistics.fastestSolvingTime, game.time)
        }
        if (p.getStatistic(Statistics.maximumPoints) < game.score) {
            p.setStatistic(Statistics.maximumPoints, game.score)
        }
    }

    private fun incrementStatistic(s: Statistics) { //TODO this should probably be in model...
        val profilesDir = context.getDir(
            context.getString(R.string.path_rel_profiles),
            Context.MODE_PRIVATE
        )
        val p = ProfileSingleton.getInstance(profilesDir, ProfileRepo(profilesDir),
                                             ProfilesListRepo(profilesDir))
        p.setStatistic(s, p.getStatistic(s) + 1)
    }
}